### DISCUSSION TITLE
Does fabric cache any textures when dealing with webgl filters?

### DISCUSSION THREAD

**Discussion by AlvesJorge** [fabricjs/fabric.js#10212] [Q&A]
Hi again,
Context:
Using Fabric's webgl filter framework I'm currently writing a class that blends two images together.
I have texture1, passed as a uniform by fabric, `uTexture`, as it's the `fabric.Image` to whom the filter is applied to.
I then have texture2, passed by me as `uMask`, which itself is a texture created from an image. 
In the end I have 3 instances of `fabric.Image`, all based on the exact same HTMLImageElement, and each of them will have a different shader will be applied to, using the `uMask` texture which is the same in all three instances too.

My question:
I noticed that I don't have to create, bind, or call `texImage2D` more than once for my texture2, and not doing so saves a ton (100s of ms) every time I don't do it. Does the same happen for the `texture1` aka the fabric.Image to which the filter is applied to? Meaning, does fabric.js create the texture1 three times, or is the texture cached? 
I can't find the creation of `uTexture` anywhere in the code :/

**Comment by asturur** [comment]
Yes fabricJS caches the texture and has a method to remove the base texture for fabricImage.
Be aware there is already a blendImage filter that blend 2 images together, you may want to look at that also to build your if it needs to be built diffrently.

Please try to follow this code from here: https://github.com/fabricjs/fabric.js/blob/master/src/filters/WebGLFilterBackend.ts#L106
And also read what the BaseFilter does: https://github.com/fabricjs/fabric.js/blob/master/src/filters/BaseFilter.ts and the BlendImage filter https://github.com/fabricjs/fabric.js/blob/master/src/filters/BlendImage.ts

If you spot ideas for improvement post them

**Comment by AlvesJorge** [comment]
> blendImage filter that blend 2 images together, you may want to look at that also to build your if it needs to be built diffrently.

My needs are a bit more specific! But still cool that this is already offered as a filter.


Thanks for the other links for `WebGLFilterBackend` and `BaseFilter` I already had built my custom filters with these in mind, but still great that you are confirming I was looking in the right place :) All is already working fine I am just looking to squeeze some more performance!

One of the "squeezes" I found was caching my own texture2.

Another one that also really sped things up was to make sure the image is already in a canvas before applying the filter. Previously I was trying to make sure the image would only be drawn in the canvas in its final state, but this (speculation) probably causes it to need to be drawn in a temporary canvas  just so it can have the filter applied to it, which really slows things down for me :) Curious to know if this makes sense to you @asturur.

**Comment by AlvesJorge** [comment]
Also curious to know if there's a way I can confirm this for myself
> Yes fabricJS caches the texture

**Reply by asturur** [reply]
There is a method 'getCachedTexture' that reference a created texture by a string that is on the fabricImage instance

**Comment by AlvesJorge** [comment]
Sorry to bother you more, but can you point me to where `uTexture` is sent as a uniform? I can't find it when I grep the code

**Reply by asturur** [reply]
I can't find it either. Could it be a generic standard name embedded for the first texture unit? uImage is clearly sent for blendImage.
Can't find info on google and i wrote this code around 7 years ago
